### PR TITLE
Claim type price

### PR DESCRIPTION
**Pull request by Holymir**
different prices for different claim types, tests fixes

### CODE REVIEW

File: contracts/Factory/IPClaimsFactory.sol [hunk 1]
```diff
@@ -81,11 +78,38 @@ contract IPClaimsFactory is IIPClaimsFactory, OwnableUpgradableImplementation {
...
+    function createClaim(bytes32 _ipfsAddress, uint8 _claimType) public returns (bool) {
+        require(_ipfsAddress != bytes32(0));
+
+        uint256 claimRecordingFeeInTokens = tokenExchangeOracle.convertUSDToTokens(claimTypesPricesInUSD[_claimType]);
+
+        if (claimRecordingFeeInTokens < ERC20(erc20Instance).allowance(msg.sender, this)) {
+            claimRecordingFeeInTokens = tokenExchangeOracle.convertUSDToTokensByLastRate(claimTypesPricesInUSD[_claimType]);
+        }
+
+        assert(erc20Instance.transferFrom(msg.sender, feeReceiver, claimRecordingFeeInTokens));
```
Comment (thcrnk) [inline contracts/Factory/IPClaimsFactory.sol]: Consider removing this assert to not lose ether for txn costs if something went wrong with limepay

File: contracts/Factory/IPClaimsFactory.sol [hunk 2]
```diff
@@ -81,11 +78,38 @@ contract IPClaimsFactory is IIPClaimsFactory, OwnableUpgradableImplementation {
...
+
+    function setSpecificClaimTypePriceInUSD(uint8 _typeID, uint256 _typePrice) public onlyOwner {
+        claimTypesPricesInUSD[_typeID] = _typePrice;
+    }
+
+    function getSpecificClaimTypePriceInUSD(uint8 _typeID) public view returns (uint256) {
+        return claimTypesPricesInUSD[_typeID];
+    }
+
+    function createClaim(bytes32 _ipfsAddress, uint8 _claimType) public returns (bool) {
```
Comment (thcrnk) [inline contracts/Factory/IPClaimsFactory.sol]: add require if the type is existing

File: contracts/Factory/IPClaimsFactory.sol [hunk 3]
```diff
@@ -81,11 +78,38 @@ contract IPClaimsFactory is IIPClaimsFactory, OwnableUpgradableImplementation {
...
-    function createClaim(bytes32 ipfsAddress) public returns (bool) {
-        require(ipfsAddress != bytes32(0));
+    function setTokenExchangeOracle(address _tokenExchangeOracle) public onlyValidAddress(_tokenExchangeOracle) onlyOwner {
+        tokenExchangeOracle = TokenExchangeOracle(_tokenExchangeOracle);
+    }
+
+    function getTokenExchangeOracle() public view returns(address) {
+        return address(tokenExchangeOracle);
+    }
+
```
Comment (thcrnk) [inline contracts/Factory/IPClaimsFactory.sol]: add getter to get length of the array with types

File: contracts/Factory/IPClaimsFactory.sol [hunk 4]
```diff
@@ -32,47 +33,43 @@ contract IPClaimsFactory is IIPClaimsFactory, OwnableUpgradableImplementation {
...
-    onlyNotEmptyAddress(_ipTokenContractAddress) {
-        ipTokenContract = ERC20(_ipTokenContractAddress);
+    function setIPTokenAddress(address _erc20InstanceAddress) public onlyOwner
+    onlyValidAddress(_erc20InstanceAddress) {
+        erc20Instance = ERC20(_erc20InstanceAddress);
     }
 
-    function getIPTokenAddress() public view returns (address _ipTokenContractAddress) {
-        return address(ipTokenContract);
+    function getIPTokenAddress() public view returns (address __erc20InstanceAddress) {
```
Comment (vladichhh) [inline contracts/Factory/IPClaimsFactory.sol]: one of the underscores could be removed :)

**Pull request merged**